_posts/2019-07-26-pysam-with-s3.md

```diff
@@ -0,0 +1,194 @@
+1) https://www.biostars.org/p/213305/
+2) https://github.com/pysam-developers/pysam/issues/557
+3) https://www.biostars.org/p/213305/ 
+
+Although several sources claim that accessing s3 objects should just work, I have not been able to find a single worked example. After many days of digging on the web and coming across several problems, my team and I figured out how to successfully use PySAM and htslib to access private S3 bucket; we describe how below.
+
+### 1) Download and install htslib version from developer branch
+   
+We believe the root of the problem is AWS's move to a [new algorithm](https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-authenticating-requests.html) for signed URL access to private objects which is not supported by the current release of htslib (v1.9); but is supported on the `devel` branch.  To use this new version, the installation instructions can be found here: https://github.com/samtools/htslib . Make sure to add the --enable-s3 option during installation. 
+
+To check whether htslib installed correctly, use the command:
+```
+which htsfile
+```
+The output should be:
+```
+/usr/local/bin/htsfile
+```
+
+### 2) Try to Run PySAM
+
+We tried to call a s3 bucket using the testing server play.min.io:3000, and uploaded a vcf file inside a test bucket
+
+##### Set environment variables
+```
+AWS_SECRET_ACCESS_KEY=zuf+tfteSlswRu7BJ86wekitnifILbZam1KYY3TG
+AWS_ACCESS_KEY_ID=Q3AM3UQ867SPQQA43P2F
+```
+##### create test.py file
+```
+from pysam import VariantFile
+
+path = "s3://play.min.io:9000/test/NA18537.vcf.gz"
+vcf = VariantFile(test2)
+for rec in vcf.fetch():
+    print(rec.pos)
+```
+
+##### And we get the error:
+```
+PermissionError: [Errno 13] could not open variant file `b's3://default@play/test/NA18537.vcf.gz'`: Permission denied
+```
+
+### 3) Successfully open file from S3 path using htslib directly
+
+The error within pysam above motivated us to test whether we can use htslib directly to call a s3 bucket. We used the testing server play.min.io:9000 and created our own test bucket called testfiles with the file NA18537.vcf.gz. 
+
+##### Create ./s3cfg file instead of using environment variables
+Since using environment variables gave us the same error, we decided to use a different method. We created ./s3cfg file in the root directory of your computer. The file should look like this:
+```
+[default]
+access_key = Q3AM3UQ867SPQQA43P2F  
+secret_key = zuf+tfteSlswRu7BJ86wekitnifILbZam1KYY3TG  
+host_base = min.io:9000
+```
+
+If you're wondering why the host_base is min.io:9000 and not play.min.io:9000, it is because htslib reads the host_base, and the file path in a specific format, and we had to play around with it for awhile to get it to read the host name in the way we wanted. In general, when calling the s3 path with htslib, the path is in the following format:
```

Comment:
When using AWS, the host-base is known to be, e.g., s3.aws.amazon.com (or whatever).